src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/DeprecationPackageMetadataCapability.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.UI.Models.Package
+{
+    internal class DeprecationPackageMetadataCapability : IDeprecationCapable
+    {
+        private readonly IPackageMetadataRetrievalAdapter _packageMetadataRetrievalAdapter;
+        private PackageDeprecationMetadataContextInfo? _deprecationMetadata;
+
+        public DeprecationPackageMetadataCapability(IPackageMetadataRetrievalAdapter packageMetadataRetrievalAdapter)
+        {
+            _packageMetadataRetrievalAdapter = packageMetadataRetrievalAdapter ?? throw new ArgumentNullException(nameof(packageMetadataRetrievalAdapter));
+        }
+
+        public AlternatePackageMetadataContextInfo? AlternatePackage => _deprecationMetadata?.AlternatePackage;
+
+        public bool IsDeprecated => _deprecationMetadata != null;
+
+        public PackageDeprecationReason PackageDeprecationReasons
+        {
+            get
+            {
+                if (_deprecationMetadata?.Reasons == null || _deprecationMetadata.Reasons.Count == 0)
+                {
+                    return PackageDeprecationReason.Unknown;
+                }
+
+                bool hasCriticalBugs = false;
+                bool hasLegacy = false;
+
+                foreach (var reason in _deprecationMetadata.Reasons)
+                {
+                    if (string.Equals(reason, PackageDeprecationReasonConstants.CriticalBugs, StringComparison.OrdinalIgnoreCase))
+                    {
+                        hasCriticalBugs = true;
+                    }
+                    else if (string.Equals(reason, PackageDeprecationReasonConstants.Legacy, StringComparison.OrdinalIgnoreCase))
+                    {
+                        hasLegacy = true;
+                    }
+                }
+
+                if (hasCriticalBugs && hasLegacy)
+                {
+                    return PackageDeprecationReason.LegacyAndCriticalBugs;
+                }
```

Comment:
We could have an earlier exit here
```suggestion
                foreach (var reason in _deprecationMetadata.Reasons)
                {
                    if (string.Equals(reason, PackageDeprecationReasonConstants.CriticalBugs, StringComparison.OrdinalIgnoreCase))
                    {
                        hasCriticalBugs = true;
                    }
                    else if (string.Equals(reason, PackageDeprecationReasonConstants.Legacy, StringComparison.OrdinalIgnoreCase))
                    {
                        hasLegacy = true;
                    }

                    if (hasCriticalBugs && hasLegacy)
                    {
                        return PackageDeprecationReason.LegacyAndCriticalBugs;
                    }
                }

                
```

Reply:
Thanks, that is a great idea. We'll do that in an isolated change in a future iteration!